server/tokenRegistryProxy.js

```diff
@@ -1,29 +1,85 @@
 require('isomorphic-fetch')
 const Sentry = require('@sentry/node')
 const {backendConfig} = require('./helpers/loadConfig')
-const cacheResults = require('./helpers/cacheResults')
+const chunk = require('./helpers/chunk')
+const Cache = require('./helpers/cache')
 
-const CACHE_TIMEOUT = 24 * 60 * 60 * 1000 // 1 day
+const MAX_SUBJECTS_PER_REQUEST = 100
+const MAX_SUBJECTS_COUNT = 2000
 
-const cachedFetch = cacheResults(CACHE_TIMEOUT)(async (url) => {
-  const response = await fetch(url)
-  if (response.status === 200) {
-    return response.json()
-  } else {
-    return null
+const cache = (() => {
+  // We expect far more tokens to not be in token registry, we keep that information in nullCache
+  // Keeping it separate helps us to not prune cached information about tokens that have real
+  // metadata in token registry
+  const cache = new Cache(1000)
+  const nullCache = new Cache(100000)
+
+  const has = (key) => cache.has(key) || nullCache.has(key)
+
+  const get = (key) => (cache.has(key) ? cache.get(key) : nullCache.get(key))
+
+  const set = (key, value) => {
+    if (value === null) {
+      nullCache.set(key, value)
+    } else {
+      cache.set(key, value)
+    }
   }
-})
+
+  return {has, get, set}
+})()
 
 module.exports = function(app, env) {
   app.post('/api/bulk/tokens/metadata', async (req, res) => {
     try {
-      const responses = await Promise.all(
-        req.body.subjects.map((subject) =>
-          cachedFetch(`${backendConfig.ADALITE_TOKEN_REGISTRY_URL}/metadata/${subject}`)
+      const subjects = req.body.subjects
+
+      if (subjects.length > MAX_SUBJECTS_COUNT) {
+        return res.json({
+          statusCode: 400,
+          Left: 'Request over max limit',
+        })
+      }
+
+      // Retrieve subjects that are cached from cache and determine
+      // which subjects need to be fetched from remote
+      const toFetch = subjects.filter((s) => !cache.has(s))
+
+      // filter both - undefined (to fetch) and null (no metadata) entries
+      const cachedTokensMetadata = subjects.map((s) => cache.get(s)).filter((s) => s != null)
```

Comment:
won't the `.filter((s) => s != null)` result in items in the null cache being ignored? I think the correct transformation should look like this:

```javascript
subjects.filter((s) => cache.has(s)).map((s) => cache.get(s))
```

Reply:
Yes that's the point, otherwise it would fill the array with null values, like `[null, null, ..., null]`. Current logic expects that subjects that don't have associated metadata are just not returned. Token registry service behaves in the same way.

Reply:
ah I see - I'd name it then cachedNonNullTokensMetadata otherwise as a reader I may be confused why the "null" metadata are not considered here and may genuinely believe it's a bug

Reply:
Fixed